File: frontend/src/app/features/auth/auth.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, computed, inject, signal } from '@angular/core';
import { Observable, tap } from 'rxjs';
import { AuthState, LoginRequest, LoginResponse } from './auth.model';

const STORAGE_KEY = 'course-eligibility-portal.auth';

@Injectable({ providedIn: 'root' })
export class AuthService {
  private readonly http = inject(HttpClient);
  private readonly state = signal<AuthState | null>(this.restore());

  readonly isAuthenticated = computed(() => {
    const session = this.state();
    return !!session && session.expiresAt > Date.now();
  });
  readonly username = computed(() => this.state()?.username ?? null);

  login(request: LoginRequest): Observable<LoginResponse> {
    return this.http
      .post<LoginResponse>('/auth/login', request)
      .pipe(tap((response) => this.setSession(response)));
  }

  logout(): void {
    this.state.set(null);
    try {
      localStorage.removeItem(STORAGE_KEY);
    } catch {
      /* storage unavailable - nothing to clean up */
    }
  }

  token(): string | null {
    const session = this.state();
    return session && session.expiresAt > Date.now() ? session.token : null;
  }

  private setSession(response: LoginResponse): void {
    const session: AuthState = {
      token: response.token,
      username: response.username,
      expiresAt: Date.now() + response.expiresIn * 1000
    };
    this.state.set(session);
    try {
      localStorage.setItem(STORAGE_KEY, JSON.stringify(session));
    } catch {
      /* private browsing / quota exceeded - session-only fallback */
    }
  }

  private restore(): AuthState | null {
    try {
      const raw = localStorage.getItem(STORAGE_KEY);
      if (!raw) {
        return null;
      }
      const parsed = JSON.parse(raw) as AuthState;
      return parsed.expiresAt > Date.now() ? parsed : null;
    } catch {
      return null;
    }
  }
}

File: backend/src/main/java/com/eligibility/portal/service/AuthService.java
package com.eligibility.portal.service;

import com.eligibility.portal.dto.request.LoginRequest;
import com.eligibility.portal.dto.response.LoginResponse;
import com.eligibility.portal.security.JwtTokenProvider;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;

@Service
public class AuthService {

    private final AuthenticationManager authenticationManager;
    private final JwtTokenProvider jwtTokenProvider;

    public AuthService(AuthenticationManager authenticationManager, JwtTokenProvider jwtTokenProvider) {
        this.authenticationManager = authenticationManager;
        this.jwtTokenProvider = jwtTokenProvider;
    }

    public LoginResponse login(LoginRequest request) {
        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(request.getUsername(), request.getPassword()));
        String token = jwtTokenProvider.generateToken(authentication.getName());
        return new LoginResponse(token, jwtTokenProvider.getExpirationMillis() / 1000, authentication.getName());
    }
}
